src/main/msp/msp.c

```diff
@@ -983,6 +984,43 @@ static bool mspProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst)
 
         break;
 
+    case MSP_MOTOR_TELEMETRY:
```

Comment:
This looks ok to me, except that blending in the ESC sensor RPM data is a double up with `MSP_ESC_SENSOR_DATA`. In general, I am wondering if we should not just append this data to `MSP_ESC_SENSOR_DATA`? Yes, it will be a misnomer then, but on a data structure level I think it will be cleaner.

Reply:
Not sure the best approach. I didn't see the `MSP_ESC_SENSOR_DATA` message because it's currently not implemented at all in the Configurator - so not sure why the message is even there or who uses it.

If we wanted to use `MSP_ESC_SENSOR_DATA` and just add the extra data it would unfortunately require a second loop. Also the message has no motor count included so any consumers would have to calculate the motor count based on dividing the total buffer by the expected data per motor (3 bytes per motor currently). This could make extending this difficult for consumers. This same technique is used for `MSP_MOTOR` so that's why I didn't add the extra elements there.

Reply:
Ah correction - it does have a motor count byte. Would still need another loop though to add the extra data.

Reply:
I don't recall the details of when `MSP_ESC_SENSOR_DATA` was added, but that was likely before the time that we set up the 'one release cycle cool down for MSP' policy, so it would have been added 'just in case' at the time that ESC sensor feedback was added.
Not sure that the extra loop has any drawbacks - we do this in other places due to the way MSP does backwards compatibility. And in terms of being economical with the remaining free MSP frame ids, I think a point can be made for not using a second one to get a different form of motor data. If we combine them, we can also leave it up to the client to use whichever source for RPM data makes more sense.

Reply:
I suppose we could use `MSP_ESC_SENSOR_DATA` for just the ESC Sensor side and have `MSP_MOTOR_TELEMETRY` for just the DSHOT bidir. But would double up the messages which are streamed continuously on the motors tab.

Reply:
Okay, I'll look at it. Will have to add a second loop that will contain DHSOT bidir RPM and packet error stats. Since both can provide RPM simultaneously we'd need both as in the Configurator we'd prefer to give DSHOT bidir data precedence as it's more realtime.

Reply:
There's a potential problem with using `MSP_ESC_SENSOR_DATA` based on how it's currently implemented:
```
#if defined(USE_ESC_SENSOR)
    case MSP_ESC_SENSOR_DATA:
        if (featureIsEnabled(FEATURE_ESC_SENSOR)) {
            sbufWriteU8(dst, getMotorCount());
            for (int i = 0; i < getMotorCount(); i++) {
                const escSensorData_t *escData = getEscSensorData(i);
                sbufWriteU8(dst, escData->temperature);
                sbufWriteU16(dst, escData->rpm);
            }
        } else {
            unsupportedCommand = true;
        }

        break;
#endif
```
It will return an "unsupported command" response if ESC Sensor is not enabled. We can't do this if we append on the DSHOT bidir data - the message will always need to return ESC Sensor data (even if all zeros). The consumers may be dependent on the message returning a "unsupported" response as a basis to determine is ESC Sensor is enabled.

In this PR explicit flags were added to the `MSP_MOTOR_CONFIG` message to indicate whether DSHOT bidir and/or ESC Sensor were enabled so it's better for the future. But that doesn't help any existing consumers of `MSP_ESC_SENSOR_DATA`.

Reply:
```
case MSP_ESC_SENSOR_DATA:
        sbufWriteU8(dst, getMotorCount());
        for (int i = 0; i < getMotorCount(); i++) {
            if (featureIsEnabled(FEATURE_ESC_SENSOR)) {
                const escSensorData_t *escData = getEscSensorData(i);
                sbufWriteU8(dst, escData->temperature);
                sbufWriteU16(dst, escData->rpm); //eRPM
            } else {
                sbufWriteU8(dst, 0);
                sbufWriteU16(dst, 0);
            }
        }
        for (unsigned i = 0; i < getMotorCount(); i++) {
            int rpm = 0;
            uint16_t invalidPct = 0;
            bool rpmDataAvailable = false;
#ifdef USE_DSHOT_TELEMETRY
            if (useDshotTelemetry) {
                rpm = (int)getDshotTelemetry(i) * 100 * 2 / motorConfig()->motorPoleCount;
                rpmDataAvailable = true;
                invalidPct = 10000; // 100.00%
#ifdef USE_DSHOT_TELEMETRY_STATS
                if (isDshotMotorTelemetryActive(i)) {
                    invalidPct = getDshotTelemetryMotorInvalidPercent(i);
                }
#endif
            }
#endif

#ifdef USE_ESC_SENSOR
            if (featureIsEnabled(FEATURE_ESC_SENSOR)) {
                escSensorData_t *escData = getEscSensorData(i);
                if (!rpmDataAvailable) {  // We want DSHOT telemetry RPM data (if available) to have precedence
                    rpm = calcEscRpm(escData->rpm);
                    rpmDataAvailable = true;
                }
            }
#endif
            sbufWriteU32(dst, (rpmDataAvailable ? rpm : 0)); //real RPM
            sbufWriteU16(dst, invalidPct);
        }
        break;
```
This would be my proposal - hope that makes everyone happy! Keeps it backwards compatible and removed the #ifdef wrapping the MSP. Follows the same logic of writing 0 if the features are not enabled.

Reply:
No, this is not really what I had in mind. There's no reason to duplicate the ESC Sensor RPM data. We'll leave the initial loop in place but make it not conditional on the ESC Sensor being compiled in and enabled. The additional DSHOT telemetry RPM and stats will be added as a second loop. Added to `MSP_MOTOR_CONFIG` will be flags that will determine if ESC Sensor and/or DSHOT telemetry are active. The consumer can use these flags to decide which it wants to use. So the only real change for legacy consumers of `MSP_ESC_SENSOR_DATA` will be that the message will always return a result rather than the lack of a response indicating that ESC Sensor is not available. Whether or not ESC Sensor is available can be explicitly determined by the flags in `MSP_MOTOR_CONFIG`.

Reply:
`escData->rpm` returns eRPM while you are calculating the "real" RPM using either `rpm = (int)getDshotTelemetry(i) * 100 * 2 / motorConfig()->motorPoleCount;` or `rpm = calcEscRpm(escData->rpm);`
Not sure if anyone would ever want to see the eRPM, but to keep it backwards compatible you would have to send it twice (one eRPM and one "real" RPM in the second loop.
I could react to it if I see the minor MSP version >=42 and then just use the "real" RPMs from the second loop. Up to you.

Reply:
@awolf78 Yes, you're correct. Are you currently displaying eRPM in your OSD? I assume so since you don't currently have access to the `motor_poles` (it's being added to `MSP_MOTOR_CONFIG` in this PR as well).

@mikeller I beginning to think using `MSP_ESC_SENSOR_DATA` for this is getting less desirable as we're just going to end up duplicating the data anyway. Alternatively if we're confident that @awolf78's application is the only one using the message we could decide to make a clean version break and redesign the message to be more appropriate. Or just go with the original clean implementation of `MSP_MOTOR_TELEMETRY` and consider `MSP_ESC_SENSOR_DATA` deprecated as it only seems to be used for this one application.

Reply:
@etracer65 I am converting it on my side from eRPM to real RPMs.
I am fine either way using `MSP_MOTOR_TELEMETRY` or `MSP_ESC_SENSOR_DATA` since this only applies if the minor version is >=42 so I am making changes on my side anyhow.

Reply:
@etracer65: I agree - it seems to be that the details like eRPM vs. RPM will not allow us to add the new data to `MSP_ESC_SENSOR_DATA` in a backwards compatible way, so going the way of adding a new frame type for the combined ESC sensor / bidirectional Dshot data, and marking `MSP_ESC_SENSOR_DATA` and the parts of the voltage / current frames sending ESC sensor data as deprecated seems to be the right way to go.
One thing to check is to make sure we have the motor pole count parameter even if bidirectional Dshot is not built, or else we won't be able to convert the ESC sensor eRPM into RPM.

Reply:
@etracer65 Why is rpm 32bit? Do those little 2 inchers exceed 0xFFFF RPM? Otherwise it would fit in uint16_t

Reply:
@awolf78 In theory yes. In reality don't know. But if you do the math with 6S, 65535 RPM is only 2600kv. I know people running 6S on little 1407 4000kv motors. Whether the ESC is really capable of the eRPM needed, the motor has the actual torque to spin the prop that fast, and whether the prop won't explode at that RPM are pesky realities. Figured better safe than sorry and not use 16bit since it's borderline.

Also don't do your redesign yet based on this PR. I still have some updates to make and add extra elements to the `MSP_MOTOR_TELEMETRY` message.

Reply:
@mikeller The `motor_poles` being included is not dependent on any defines. The parameter is available and I added it to the MSP message regardless of `USE_ESC_SENSOR` and `USE_DSHOT_TELEMETRY`.